### PR TITLE
implement byte-aware chunking

### PR DESCRIPTION
**Pull request by dezren39**
- Add chunkType parameter, allowing multi-byte aware chunking.
   - Third parameter, default behavior if excluded is equal to previous behavior.
   - "-1"  : same as previous default behaviour. every character = 1
   - "0"   : every character = as many bytes as the character is
   - ">0" : every character = as many bytes as the character is, up to N
     - This is the use case I need, as I need each emoji to count as "2".
- Some characters may encode to more bytes than expected.
- Added new tests to accommodate changes and verify backwards compatibility.
  - Made some subjective changes to pre-existing tests to better allow them to be read in a variety of terminals/fonts.
     - When working with emoji, I've had issues where they display differently everywhere.
     - Parsing the variable names has been easier overall, while sometimes more annoying on properly encoding editors.
- Changed chunkSize validation to use Numbers.parseInt and Numbers.isNaN
  - It has better overall compatibility, though does allow weird things like `chunk("text", "2 years", "-1 years")`
- To allow easier compatibility with shell, etc:
  -  npm run-scripts: start, chunk, chunk-text
  -  bin/server.js
  -  package.json bin property
- Added keywords to package.json
- Updated dependencies.
   - Had a hard time parsing between needed dev dependencies.
   - Some _may_ be able to be removed.
- Worked towards ESM-native code with CJS backwards compatibility.
  - Switched code/tests to imports.
  - package.json module, exports, properties.
  - CJS require still compatible.
  - Local build/dev may have new engine compatibility limitations.
  - I was unable to actually set "type":"module", so I clarified package.json property: "type":"commonJs"
    - The issue I had was jest and babel had different compatibility issues, their solutions seemed to conflict.
    - None of the other PRs related to the errors I found had any working solutions.
    - Jest devs didn't get community feedback in 2019 to prioritize ES support.
    - I considered other solutions, but already had made enough major changes for one PR.
       - On this note, I know algolia uses this for a lot of processing or something. I haven't performance tested anything and literally have just seen it "work on my machine". I made the changes for my own needs, but in order to get them mainlined can definitely rip out anything like imports instead of requires or something if it better aligns with something I'm missing. For instance, I updated a bunch of babel things and maybe you had a reason for locking it where it was outside of this module's requirements. 
- Incremented minor version, because although adding a parameter isn't a breaking change --
   - I do worry the overall changes may have some possibility of compatibility issue. Babel helps alleviate this concern though.
   - The effort to allow this capability has increased the complexity of this project.
      - I've tried to add enough testing to mitigate for the risks this presents.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,45 +1,127 @@
...
+  ) {
+    throw new TypeError(
+      'Type should be provided as 3rd (optional) argument and parseInt to a value >= -1.'
+    );
+  }
+};
 
-  while (characters.length > chunkSize) {
-    const splitAt = characters.lastIndexOf(' ', chunkSize);
+const chunkLength = (characters, chunkType) => {
```
Comment (Haroenv) [inline src/index.js:36]: there's a lot of syntactic changes, which makes it harder to review the change itself, is it possible to split that up in separate PRs, one for the syntax changes, and one for the functional change?

File: src/index.js [hunk 2]
```diff
@@ -1,45 +1,127 @@
...
-  assertIsValidText(text);
-  assertIsValidChunkSize(chunkSize);
-
-  const chunks = [];
-  let characters = runes(text);
+const assertIsValidChunkType = (
+  chunkType,
+  chunkTypeParseIntNaN,
+  chunkTypeInt
+) => {
```
Comment (Haroenv) [inline src/index.js:23]:
```suggestion
function assertIsValidChunkType (
  chunkType,
  chunkTypeParseIntNaN,
  chunkTypeInt
) {
```

I don't think for example that everything should be an arrow function like this

**Review by Haroenv — COMMENTED** [review]
the change looks very promising, but it's hard to review like this

**Comment by dezren39** [comment]
Sorry!

Happy to cut it down, there are a few more commits in my local repo for context but not super useful ones which is why I squashed into one big one.

I think all the import statements, package and dependency changes could be removed except a devDependency needed for jest to use TextEncoder. I'd have to test that.

Likewise the bin file and package changes for that could maybe be their own change.

Not sure if I can cut down or clarify the tests / main code changes further, but I can try. I didn't see any obvious redundant or useless code by the time I'd pushed, but who knows if I look it over again. 🤔😅

File: src/index.js (hunk 1 above)
Comment (dezren39) [reply src/index.js:36]: Probably yeah! Haven't tried but should be easy to switch back. I saw your other suggestion to switch some back to regular functions, that make sense and can do, is there any specific time to prefer arrow function vs regular function? Should I port all of them back to regular functions?

**Comment by Haroenv** [comment]
I mostly meant to have a separate PR for your changes related to syntax, and then you can have this PR for just the code change, this way we will be able to review just the relevant code

**Comment by dezren39** [comment]
closing, will open new one with new branch

**Comment by dezren39** [comment]
made #8 for the chunkType changes in addition to #9 for the syntax/etc changes
